lib/hsa/hsaPulsarUpdatePhase.cpp

```diff
@@ -129,6 +109,36 @@ int hsaPulsarUpdatePhase::wait_on_precondition(int gpu_frame_id) {
         return -1;
     metadata_buffer_precondition_id =
         (metadata_buffer_precondition_id + 1) % metadata_buf->num_frames;
+
+
+    // Wait for new gain
+    DEBUG("Waiting for gain_buf_id={:d} to be full; gpu_frame_id={:d}", gain_buf_id,
+          gpu_frame_id);
+    if (first_pass) {
+        uint8_t* frame =
+	  wait_for_full_frame(gain_buf, unique_name.c_str(), gain_buf_id);
+        host_gain = (float*)gain_buf->frames[gain_buf_id];
+	update_phase = true;
+        mark_frame_empty(gain_buf, unique_name.c_str(), gain_buf_id);
+	gain_buf_id = (gain_buf_id + 1) % gain_buf->num_frames;		
+        if (frame == NULL)
+            return -1;
+    } else {
+      auto timeout = double_to_ts(0);
+      int status = wait_for_full_frame_timeout(gain_buf, unique_name.c_str(),
+					       gain_buf_id, timeout);
+      DEBUG("status of gain_buf_id[{:d}]={:d} ==(0=ready 1=not)",
+	    gain_buf_id, status);
+      if (status == 0) {
```

Comment:
You should add `std::lock_guard<std::mutex> lock(_pulsar_lock);` after the `if (status ==0)` so that you don't update the gain at the same time as a phase calculation is happening.